File: october_2024/potd-15-10-2024.py
"""
white: This keeps track of how many '1's have been encountered as we iterate through the string. It starts at 0.
res: This accumulates the number of steps needed. It also starts at 0.
j: This is the index used to traverse the string.
n: This stores the length of the input string s

For example, if s = "001101":

    Start with white = 0, res = 0.
    At index 0: s[0] = '0', res += 0 (still 0). white = 0.
    At index 1: s[1] = '0', res += 0 (still 0). white = 0.
    At index 2: s[2] = '1', increment white to 1.
    At index 3: s[3] = '1', increment white to 2.
    At index 4: s[4] = '0', add white (2) to res (now res = 2).
    At index 5: s[5] = '1', increment white to 3.

Final res is 2, which represents the steps needed to convert all '0's.
"""
class Solution:
    def minimumSteps(self, s: str) -> int:
        white=0
        res=0
        j=0
        n=len(s)
        while(j<n):
            if s[j]=='0':
                res+=white
            white+=1 if s[j]=='1' else 0
            j+=1
            
            
        return res

File: october_2024/potd_15_10_2024.cpp
// Traversal: We traverse the string from right to left. For every '1' encountered, we add the number of '0's seen so far (in count) to the total ans because each '1' has to move past those '0's.

class Solution {
public:
    // Function to calculate the minimum number of steps required
    // based on the string 's' containing '0's and '1's.
    long long minimumSteps(string s) 
    {
        long long ans = 0;  // Variable to store the final answer (minimum steps).
        
        // If the string is empty or has only one character, no steps are needed.
        if(s.length() == 0 || s.length() == 1)
            return 0;
        
        int r = s.length() - 1;  // Set 'r' to the index of the last character in the string.
        long long count = 0LL;   // 'count' keeps track of the number of '0's encountered.
        
        // Traverse the string from the last character to the first.
        while(r >= 0)
        {
            // If the current character is '1', we add the current count of '0's to 'ans'.
            // This is because each '1' should move past all the '0's seen so far.
            if(s[r] == '1')
                ans += count;
            else
                // If the current character is '0', increment the 'count' of '0's.
                count += 1LL;
            
            r--;  // Move to the previous character in the string.
        }
        
        // Return the total number of steps (ans).
        return ans;
    }
};

// Main function to test the minimumSteps function
int main() {
    Solution solution;

    // Example input string
    string s = "1010010001";

    // Call the function and store the result
    long long result = solution.minimumSteps(s);

    // Output the result
    cout << "Minimum number of steps required: " << result << endl;  // Output should be 5

    return 0;
}

// Expected Output : Minimum number of steps required: 14
